donkeycar/parts/tub_v2.py

```diff
@@ -120,13 +124,48 @@ def run(self, *args):
         self.tub.write_record(record)
         return self.tub.manifest.current_index
 
+    def run_threaded(self, *args):
+        assert len(self.tub.inputs) == len(args), \
+            f'Expected {len(self.tub.inputs)} inputs but received {len(args)}'
+        record = dict(zip(self.tub.inputs, args))
+        # push the record into queue
+        self.record_queue.put(record)
+        self.mutex.acquire()
+        count = self.tub.manifest.current_index
+        self.mutex.release()
+        return count
+
+    def update(self):
+        self.on = True
+        while self.on or not self.record_queue.empty() > 0:
+            try:
+                # pop a record
+                record = self.record_queue.get()
+                # write the record to disk
+                self.mutex.acquire()
+                self.tub.write_record(record)
+                self.mutex.release()
+                # notify this task is done
+                self.record_queue.task_done()
+            except Empty:
+                pass
+
+    """
+    Let's disable this first
     def __iter__(self):
         return self.tub.__iter__()
+    """
 
     def close(self):
+        self.mutex.acquire()
```

Comment:
The `queue.join()` should happen before we close the `tub` because this will close the files. I don't think you need a lock here, either.

Reply:
Calling `queue.join()` before tub is closed sounds good.  For the lock, here I added one because at the time `self.tub.close()` is called, there could be the worker thread still trying to access `self.tube.write_record`, and `Tub` class has no way to protect those two events happen one after another.

Reply:
Sorry, i missed the call to `join()`. That should do the right thing here.